### PR TITLE
[Painless] Add extensive tests for def to primitive casts

### PR DESCRIPTION
**Pull request by jdconrad**
This adds tests for each possible cast of def to a primitive type both implicit and explicit.  This also fixes a minor bug where we were only checking the type of a def to be Number in some explicit casts.  This does not work because it allows possible unintended casts from BigInteger and BigDecimal to primitive types since they both extend Number but are not included as part of the Painless casting model.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java [hunk 1]
```diff
@@ -696,64 +725,137 @@ public static double defTodoubleImplicit(final Object value) {
...
+                value instanceof Short   ||
+                value instanceof Integer ||
+                value instanceof Long    ||
+                value instanceof Float   ||
+                value instanceof Double
+        ) {
             return ((Number)value).byteValue();
+        } else {
+            throw new ClassCastException("cannot explicitly cast " +
+                    "def [" + PainlessLookupUtility.typeToUnboxedType(value.getClass()).getCanonicalName() + "] to byte");
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java]: Aren't all the boxed types handled above? when would we hit this else case? I just wonder about using typoToUnboxedType on a non-boxed type?

**Review by rjernst — APPROVED** [review]
Looks good, just one question.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java]: Good point :).  Will fix.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review!  Will commit as soon as CI completes.

**Pull request merged**